File: 322. Coin Change/322.cpp
// n = coins.size() m = amount
// time: O(n*m)
// space: O(m)
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        // array that stores the number of coins to make up amount = index
        // not possible to achieve amount with amount + 1 coins so set that as max value
        vector<int> numcoins(amount + 1, amount + 1);
        //initialize
        numcoins[0] = 0;

        // dynamic programming: latter combinations might yield better answer
        for (int coin: coins){
            // infinite coins: start from coin to amount, can accumulate
            // note: limited coins start backwards
            for (int i = coin; i <= amount; i++){
                numcoins[i] = min(numcoins[i], numcoins[i-coin] + 1);
            }
        }

        if (numcoins[amount] == amount + 1)
            return -1;
        else
            return numcoins[amount];
    }
};

File: 322. Coin Change/322.py
"""
dynamic programming (bottom up)
n = len(coins)
m = amount
time: O(mn)
space:O(m)
"""
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float('inf')] * (amount + 1)
        dp[0] = 0

        for coin in coins:
            for i in range(1, len(dp)):
                if (i-coin) >=0 and (i-coin) <= amount:
                    dp[i] = min(dp[i], dp[i-coin] + 1)

        # print(dp)
        if dp[amount] == float('inf'):
            return -1
        else:
            return dp[amount]
"""
clarifications
- coin[i] > 0
- amount >= 0 

dynamic programming
dp should have size of amount + 1: i ranges from 0 to amount
-> handle out of bound edge cases
dp[i] = minimum number of coins I need to make up amount i

initialization of dp
dp[i] = float('inf')
dp[0] = 0

for each coin:
    for each index i from smallest to larget in dp where i > 0:
        dp[i] = min(dp[i], dp[i-coin] + 1)

return dp[amount] -> check if infinity then return -1 instead
"""
